lib/puppet/provider/sysctl/augeas.rb

```diff
@@ -39,52 +39,73 @@ def self.sysctl_get(key)
 
   confine :feature => :augeas
 
-  def self.instances(reference_resource = nil)
+  def self.instances(reference_resources = nil)
     return @resource_cache if @resource_cache
 
-    resources = nil
+    resources = []
+    reference_resource_titles = reference_resources.map { |_ref_name, ref_obj| ref_obj.title }
+
+    sysctl_output = ''
+    if reference_resources
+      resource_dup = reference_resources.first.last.dup
+      files = ['/etc/sysctl.conf', resource_dup[:target]].compact
+
+      files.each do |file|
+        resource_dup[:target] = file
+        augopen(resource_dup) do |aug|
+          resources ||= []
+
+          reference_resource_titles.each do |ref_title|
+	    value = aug.get("$target/#{ref_title}")
+
+	    if value
+              resource = {
+                :name    => ref_title,
+                :ensure  => :present,
+                :persist => :true,
+	        :value   => value
+              }
+
+              # Only match comments immediately before the entry and prefixed with
+              # the sysctl name
+              cmtnode = aug.match("$target/#comment[following-sibling::*[1][self::#{ref_title}]]")
+              unless cmtnode.empty?
+                comment = aug.get(cmtnode[0])
+                if comment.match(/#{resource[:name]}:/)
+                  resource[:comment] = comment.sub(/^#{resource[:name]}:\s*/, "")
+                end
+              end
+
+              resources << resource
+            end
+	  end
```

Comment:
The indentation here is a bit weird, as it's aligned with the line below.

Reply:
Weird, I'll update that.